igvm/src/c_api.rs

```diff
@@ -0,0 +1,763 @@
+            }
+        }
+        if result == IgvmApiError::IGVMAPI_OK {
+            // SAFETY: The fetching and update of the global variable is safe because
+            // it is using an Atomic type.
+            let handle = unsafe { IGVM_HANDLE_FACTORY.fetch_add(1, Ordering::Relaxed) };
+            igvm.buffers.insert(handle, header_binary);
+            handle
+        } else {
+            result.0
+        }
+    } else {
+        IgvmApiError::IGVMAPI_INVALID_HANDLE.0
+    }
+}
+
+/// Prepare a buffer containing the associated file data in binary form for the
+/// entry with the given index for a particular section in a previously parsed
+/// IGVM file.
+///
+/// The buffer containing the data is returned via a handle from this function.
+/// The handle can be used to access a raw pointer to the data and to query its
+/// size. The buffer handle remains valid until it is closed with a call to
+/// [`igvm_free_buffer()`] or the parsed file handle is closed with a call to
+/// [`igvm_free()`].
+///
+/// If any error occurs then an [`IgvmApiError`] value is returned instead of a
+/// handle. Handles are always greater than zero so a return value greater than
+/// zero indicates a valid handle has been returned.
+#[no_mangle]
+pub extern "C" fn igvm_get_header_data(
+    handle: IgvmHandle,
+    section: IgvmHeaderSection,
+    index: u32,
+) -> IgvmResult {
+    // SAFETY: Safe access to the IGVM_HANDLES map through a mutex.
+    let mut lock = unsafe { IGVM_HANDLES.lock().unwrap() };
+    let igvm_handle = lock.get_mut(&handle);
+
+    if let Some(igvm) = igvm_handle {
+        match section {
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => {
+                if let Some(header) = igvm.file.directive_headers.get(index as usize) {
+                    let mut file_data = Vec::<u8>::new();
+                    let _result =
+                        header.write_binary_header(0, &mut Vec::<u8>::new(), &mut file_data);
+                    if !file_data.is_empty() {
+                        // SAFETY: The fetching and update of the global variable is safe because
+                        // it is using an Atomic type.
+                        let handle = unsafe { IGVM_HANDLE_FACTORY.fetch_add(1, Ordering::Relaxed) };
+                        igvm.buffers.insert(handle, file_data);
+                        handle
+                    } else {
+                        IgvmApiError::IGVMAPI_NO_DATA.0
+                    }
+                } else {
+                    IgvmApiError::IGVMAPI_NO_DATA.0
+                }
+            }
+            IgvmHeaderSection::HEADER_SECTION_PLATFORM => IgvmApiError::IGVMAPI_NO_DATA.0,
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => IgvmApiError::IGVMAPI_NO_DATA.0,
+            _ => IgvmApiError::IGVMAPI_NO_DATA.0,
+        }
+    } else {
+        IgvmApiError::IGVMAPI_INVALID_HANDLE.0
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::undocumented_unsafe_blocks)]
```

Comment:
Lets enforce SAFTEY comments on test code too, since test code is real too. Since if the test code isn't right, then we can't really trust it.